### PR TITLE
Freewater modelling in python for mouse imaging

### PR DESCRIPTION
**Pull request by mdesco**
@daducci, maybe you can valide the .m file. I thought I did not change it but it seems so...

There is still the part with AA(2:end) that we can change...

Let me know what you think

### CODE REVIEW

File: python/amico/core.py [hunk 1]
```diff
@@ -292,7 +308,23 @@ def fit( self ) :
                         y_est = np.dot( A, x )
                         den = np.sum(y**2)
                         NRMSE[ix,iy,iz] = np.sqrt( np.sum((y-y_est)**2) / den ) if den > 1e-16 else 0
-
+                                            
+                    if self.get_config('doSaveCorrectedDWI') :
```
Comment (daducci) [inline python/amico/core.py]: Can we generalize this "correction" also to other models? Does it make sense?
This part seems to be related only to the WaterFree model and I think it breaks with other models (e.g. self.model.d_isos might not be in the model). If so, perhaps we should find a way to save "extra parameters/maps" that are specific to each model.

File: python/amico/core.py [hunk 2]
```diff
@@ -51,8 +52,9 @@ def __init__( self, study_path, subject ) :
 
         self.set_config('peaks_filename', None)
         self.set_config('doNormalizeSignal', True)
+        self.set_config('doKeepb0Intact', False)  # does change b0 images in the predicted signal
         self.set_config('doComputeNRMSE', False)
-
+        self.set_config('doSaveCorrectedDWI', False)
```
Comment (daducci) [inline python/amico/core.py]: I like the idea to save some extra data like e.g. the predicted dMRI signal or the estimated coefficients. I thinbk we just need to find a way to be more general, as every model might need special care.

Comment (mdesco) [inline python/amico/core.py]: I agree. Right now, this is just a hack...

Comment (daducci) [inline python/amico/core.py]: I'm ok with the, the whole project (especially the python version) is still work-in-progress. We just need to be sure that it does not break the processing with other models, like referring to fields that are specific to the WaterFree model

Comment (mdesco) [inline python/amico/core.py]: I don't think it would break other models, right? I'm just using them in the Freewater part...

Comment (daducci) [inline python/amico/core.py]: I saw some modification in `core.py` as well:

- [Lines 272-274](https://github.com/mdesco/AMICO/blob/0625d984790cd79b57efac3c471a79aa70ff3b68/python/amico/core.py#L272-L274)
- [Line 313](https://github.com/mdesco/AMICO/blob/0625d984790cd79b57efac3c471a79aa70ff3b68/python/amico/core.py#L313)

The parameters `d_isos` might not be present in others models, e.g. NODDI. Or perhaps I am messing up while looking at the diff file (my first pull request..)

Comment (mdesco) [inline python/amico/core.py]: you are right. this should not be there... let me think about it. We can sit down together and fix this.

Comment (mdesco) [inline python/amico/core.py]: I have a couple of small modifications to make sure other models did not get broken... I tested with NODDI and FreeWater and seems to work fine... 

You might want to have another go at it.

**Comment by daducci** [comment]
I think it looks fine. As you said, it's a workaround for now. Later we need to find a way to move all this processing inside the models, as each can have different requirements. We'll see together if you want (I have some ideas). but for now I think we can merge. Ok with you?

**Comment by mdesco** [comment]
Good for me. Your the boss!

**Comment by robinglord** [comment]
Hey @mdesco,
First of all, thanks for this python implementation of the Free-water model.
I'm trying to understand the science behind it, is it inspired from the 2009 Pasternak article (Free Water Elimination and Mapping from Diffusion MRI) ? If not, could you  please provide any of the references you used to make this implementation?
Thanks a lot for your help!

**Pull request merged**